### PR TITLE
Run end2end tests on PRs

### PR DESCRIPTION
**Pull request by EngHabu**
# TL;DR
This is an initial release that enables End2end tests on any PR as well as merges to master. It's also possible for any fork to run the same CI steps.

This adds two additional github actions steps:
- Setup a [KinD](https://kind.sigs.k8s.io/) cluster (Kubernetes in Docker)
- Run `end2end_execute`:
   - Clone flyte repo master
   - Replace propeller image with git sha's image
   - Run `make end2end_execute` from flyte repo. This deploys `flyte` manifests and run `endtoend.yaml` pod and waits for all workflows to finish executing as expected.

## Type
 - [ ] Bug Fix
 - [X] Feature
 - [ ] Plugin

## Are all requirements met?

 - [X] Code completed
 - [X] Smoke tested
 - [X] Unit tests added
 - [X] Code documentation added
 - [X] Any pending items have an associated Issue

## Tracking Issue
https://github.com/lyft/flyte/issues/324

### CODE REVIEW

**Comment by kumare3** [comment]
Cool lgtm

File: .github/workflows/pull_request.yml [hunk 1]
```diff
@@ -17,3 +21,14 @@ jobs:
...
+      - name: End2End
+        env:
+          DOCKER_USERNAME: ${{ github.actor }}
+          DOCKER_PASSWORD: ${{ secrets.enghabu_secret }}
+        run: |
+          kubectl cluster-info
+          kubectl get pods -n kube-system
+          echo "current-context:" $(kubectl config current-context)
+          echo "environment-kubeconfig:" ${KUBECONFIG}
+          PROPELLER=${{ secrets.flytegithub_repo }}/operator:${{ github.sha }} make end2end_execute
```
Comment (kumare3) [inline .github/workflows/pull_request.yml:34]: config changes will cause problems?

File: boilerplate/lyft/end2end/end2end.sh [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+# TODO: load all images
+echo "Loading github docker images into 'kind' cluster to workaround this issue: https://github.com/containerd/containerd/issues/3291#issuecomment-631746985"
+docker login --username ${DOCKER_USERNAME} --password ${DOCKER_PASSWORD} docker.pkg.github.com
+docker pull docker.pkg.github.com/${PROPELLER}
+kind load docker-image docker.pkg.github.com/${PROPELLER}
+
+pushd ${OUT}
+# TODO: Only replace propeller if it's passed in
+# TODO: Support replacing other images too
+sed -i.bak -e "s_docker.io/lyft/flytepropeller:.*_docker.pkg.github.com/${PROPELLER}_g" ${OUT}/kustomize/base/propeller/deployment.yaml
```
Comment (wild-endeavor) [inline boilerplate/lyft/end2end/end2end.sh:25]: I think eventually we should support something like `envsubst`

File: .github/workflows/master.yml [hunk 3]
```diff
@@ -29,3 +29,14 @@ jobs:
           push_git_tag: true
           registry: docker.pkg.github.com
           build_extra_args: "--compress=true"
+      - uses: engineerd/setup-kind@v0.4.0
+      - name: End2End
+        env:
+          DOCKER_USERNAME: ${{ github.actor }}
+          DOCKER_PASSWORD: ${{ secrets.enghabu_secret }}
```
[the remark this answers is not in the document]
Comment (EngHabu) [reply .github/workflows/master.yml]: reverted
